Stats page

nzoba opened the pull request:
Since I only use J2K I added it there. I thought of opening a PR to main first but I don't know a thing about compose so I would end up using xml (pretty sure it wouldn't be accepted that way and don't want to do twice the work with the conversion to compose tbh).
If you are not interested in this feature or don't want to add it before main, feel free to close the PR, I don't mind. But well it's mostly additions, with very few changes so nothing existing should break.

Almost all the stats mentionned in [the main issue](https://github.com/tachiyomiorg/tachiyomi/issues/7213) should be here (at least the ones we can do as of now)

The design (and base code) is heavily inspired from [AL-chan](https://github.com/zend10/AL-chan)

I added 2 pages:
- **Stats page**: Some stats numbers + 1 status pie chart + 1 score bar chart. 

| Overflow dialog | Stats page   |
| ----- | ----- |
| ![Overflow dialog](https://user-images.githubusercontent.com/55888232/182032071-e387611f-2bb2-4cd3-8c93-1a7ac2053d79.png) | ![Stats page](https://user-images.githubusercontent.com/55888232/182029119-da075bd3-746d-47fc-b1e1-2e16bb60b235.png) |

- **Stats details page**: 10 different stats (chart for each one except source and tag). Can be filtered with the chips and sorted (count, chapters read, mean score). When you have multiple trackers (= multiple scores for one manga), the manga score is the average of all scores converted to 10 points

| Available stats  | Serie type stat   |
| ----- | ----- |
| ![Available stats](https://user-images.githubusercontent.com/55888232/182029274-8aa864c2-f724-460f-87d2-f4601bc7a29f.png) | ![Serie type chart](https://user-images.githubusercontent.com/55888232/182029348-19e02ac0-8167-478d-b1e9-082546ce6970.png) |

| Status stat with filters  | Length stat  |
| ----- | ----- |
| ![Status stat](https://user-images.githubusercontent.com/55888232/182029585-ebec93c0-c168-45a8-8546-71f3768a6b84.png) | ![Length stat](https://user-images.githubusercontent.com/55888232/182029483-8345c62f-bd0e-464c-9508-e50263b32991.png) |

PS: My bad for the enormous PR but I can't really split it (or the other PRs wouldn't work without the first one). Well it's a lot of duplicated code between all the stats so it's not really 2700 different lines

Jays2Kings commented:
This actually would've been a good screen to start using compose on, but since I haven't implemented it in the main branch I'll take this. 

I'll review this all at a later time though

nzoba commented:
Oh yeah that's right, could have been a good idea. Feel free to change it to compose if you want, I'm not sure where to start if I do it myself

Sure take your time, anyway I won't have time to process your feedback for the next 2 weeks :)

Btw I added a progress bar when switching between stats but it's never showing, probably hidden by a chart or something, could have been useful especially with large libraries or tags (even thought I limited the number for these). If you have any idea how to fix that after reviewing, I'm all ears

MrChuw commented:
You could also add the reading time (not counting if the person, for example, spends 1 minute on the same part).

The option to save statuses in backup.

And I hope it will be added soon.

nzoba commented:
> You could also add the reading time (not counting if the person, for example, spends 1 minute on the same part).

It's planned, it was not possible to do it when I opened this PR but a [PR](https://github.com/Jays2Kings/tachiyomiJ2K/pull/1360) to add it was merged today, so I will add the reading time stats when I find the time

> The option to save statuses in backup.

Well the stats depend of your manga, history etc... So if you backup everything, the stats will stay the same.
Likewise, if you remove a manga from your library, you will also lose the stats related to it, at least for now

MrChuw commented:
I think it would be interesting to put it to save in the db and not just use the manga in the library because then the person would have information about everything they already read and not just what is in the library.

In my case I keep the ones I finished and dropped in the library, but there are people who might not keep them.



I don't know how to quote another comment on mobile.

nzoba commented:
I agree it's not perfect for the stats but this is how the history works. Btw what I said earlier is not totally true, you don't lose a manga's history directly after removing it from library, but only if you clear database data in the advanced settings, same for the stats. Not sure if the backup saves the removed manga data, though.
Feel free to open an issue if you want to request changes

But even if it would be useful, I won't save the stats for now, it would be far too much work (and data) and a better way would probably be to keep removed library manga's data (even if not displayed) or something similar
But I won't take this kind of decision, it's up to Jay or the developers of the main app (if I or someone else adds the stats to it)

MrChuw commented:
I'm the type that prefers as much stuff to be in the backup, I have 300 manga and it takes up 0.98mb and honestly I wouldn't care if it was taking up 50mb, but I know most aren't like that. And I would like all status to be in backup.

I understand your side, who knows in the future I open a PR. Thanks for explaining better

One last thing, did they follow the chosen theme in the app?

nzoba commented:
> I'm the type that prefers as much stuff to be in the backup, I have 300 manga and it takes up 0.98mb and honestly I wouldn't care if it was taking up 50mb, but I know most aren't like that. And I would like all status to be in backup.

Well the biggest work would be to save the stats in the db, for now everything is calculated and nothing is saved. But most stats only concern the current library manga (the only exception being read duration probably).
So if you backup everything else (manga, history etc...) then all the stats will be recalculated without any missing after recovering a backup, except read duration
So I guess it could be reduced to only saving read duration stats but it would still need quite some work and would involve modifying the db schema, which is something I want to avoid if possible

> One last thing, did they follow the chosen theme in the app?

The charts colors no, but everything else yeah (texts color, cards, background etc...)

jferdous70 commented:
for backing up manga not in library, it could be copied from TachiSy

https://github.com/jobobby04/TachiyomiSY/commit/e633da5567e6cc33beff1757d50c55b7e729a644

nzoba commented:
So I added the read duration stats. Here are a few screenshots:

I replaced the category stat number by the read duration one on the stats page.
I changed the setup of the details a bit to add the read duration everywhere

| Stats page | Serie type stat |
| ----- | ----- |
![Stats page](https://user-images.githubusercontent.com/55888232/185948283-e11c5a12-97d1-44de-844b-9f873499fe09.png) | ![Serie type stat](https://user-images.githubusercontent.com/55888232/185948354-72317b91-4712-4449-9cb8-546d934cce84.png) |

I added a new read duration stat. By default it displays a chart (the first day of the week depends of your locale) and the week data. If you click on a bar (= day) only the info of the selected day will be displayed. You can change the week by either clicking on an arrow or on the date/range (it will open a date picker)

| Week read duration | Day read duration |
| ----- | ----- |
| ![Week read duration](https://user-images.githubusercontent.com/55888232/185948625-7e625b40-d8bb-47b3-ad79-a7ba2cf08e41.png) | ![Day read duration](https://user-images.githubusercontent.com/55888232/185948809-85c4ece2-8ef1-46e6-ae86-bc0e8c03ebe8.png) |

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/ui/more/stats/StatsPresenter.kt:
```diff
@@ -0,0 +1,86 @@
+            .filter { includedCategories.isEmpty() || it.value.any { manga -> manga.category in includedCategories } }
+            .filterNot {
+                val manga = it.value.first()
+                MANGA_NON_COMPLETED in restrictions && manga.status == SManga.COMPLETED ||
+                    MANGA_HAS_UNREAD in restrictions && manga.unread != 0 ||
+                    MANGA_NON_READ in restrictions && manga.totalChapters > 0 && !manga.hasRead
```
I'd wrap each && logic in these with parenthesis, readability is clear but i dont trust kotlin sometimes without them

Jays2Kings commented on app/src/main/res/values/strings.xml:
```diff
@@ -972,6 +972,41 @@
+    <string name="count">Count</string>
+    <string name="not_rated">Not Rated</string>
+    <string name="source">Source</string>
+    <string name="serie_type">Serie Type</string>
```
this should be series type right? series type already exists

Saud-97 replied:
I think so too. Serie isn't an English word afaik.

nzoba replied:
Oh yeah my bad, both words exist in my primary language (french)

Jays2Kings reviewed (commented):
Honestly I dont think I'll go through this whole pr, since most of these are additions not changes.

Once I merge I'll see if theres some strings we can reuse because its a lot right now

Jays2Kings commented on app/src/main/res/values/strings.xml:
```diff
@@ -972,6 +972,41 @@
+    <string name="tracker">Tracker</string>
+    <string name="length">Length</string>
+    <string name="start_year">Start Year</string>
+    <string name="serie_types_">Serie Types: %1$d</string>
+    <string name="sources_">Sources: %1$d</string>
+    <string name="status_">Status: %1$d</string>
+    <string name="languages_">Languages: %1$d</string>
+    <string name="categories_">Categories: %1$d</string>
+    <string name="sort_">Sort: %1$s</string>
```
it might make sense to use a compound drawable for these in those dropdown (if possible)
* idk what to use for series type rn
* source can use the browse icon
* status you can use the same as the group by icon
* langauge can use the clipa art google translate icon
* categories can use the label icon
* sort uses sort icon

nzoba replied:
Good idea, I'll look into that

Saud-97 commented:
I suggest changing the "Stat: string". It will look better if "Stat:" was omitted or moved elsewhere.

![image](https://user-images.githubusercontent.com/39028181/186004568-89283bee-f95b-43b2-93c8-4a2e3b812d1b.png)
![image](https://user-images.githubusercontent.com/39028181/186004571-9336cdb9-30e7-4e87-b453-8e120bddca37.png)

nzoba commented:
> Honestly I dont think I'll go through this whole pr, since most of these are additions not changes.

Yeah my bad for the PR size. And as you said I don't think I break anything existing

> Once I merge I'll see if theres some strings we can reuse because its a lot right now

I tried to reuse strings as much as possible but I may have forgotten some

nzoba commented:
So I added the icons, here is what it looks like

Let's cross fingers there won't be much bugs with this much lines

![Stats chips](https://user-images.githubusercontent.com/55888232/186490223-56fcab19-2b8b-48b6-a824-c9230511e4d3.png)

AntsyLich commented:
Time to yoink this for main

nzoba commented:
Yeah, I can adapt the code to make it work in main (since some things don't exist there like series type) but what is blocking me is compose, I have no idea how to convert it so I would need to learn it first and I'm kinda busy these days

Jays2Kings commented:
The benefits of waiting to switch to compose

although when it comes time to convert this...

garvongithub commented:
Hey @nzoba big fan of your work, I really like the stats page and all other things you have added. I just had one question. How is mean score in stats calculated? Does it take in account all of the tracked enteries (even the ones that are not scored by the user) or only the ones that are scored by the user?

nzoba commented:
It only takes the ones where the score is superior to 0. It calculates the mean score for each title that has at least one scored tracker (if you have multiple trackers with different scores for one title, it takes the average). Then it also calculates the average of all the titles with a mean score
